go/consensus/tendermint/apps/governance/state/state.go

```diff
@@ -0,0 +1,358 @@
+}
+
+// Proposal looks up a proposal by its identifier.
+func (s *ImmutableState) Proposal(ctx context.Context, id uint64) (*governance.Proposal, error) {
+	proposalRaw, err := s.getProposalRaw(ctx, id)
+	if err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	if proposalRaw == nil {
+		return nil, governance.ErrNoSuchProposal
+	}
+	var proposal governance.Proposal
+	if err = cbor.Unmarshal(proposalRaw, &proposal); err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	return &proposal, nil
+}
+
+// Votes looks up all votes for a proposal.
+func (s *ImmutableState) Votes(ctx context.Context, id uint64) ([]*governance.VoteEntry, error) {
+	it := s.is.NewIterator(ctx)
+	defer it.Close()
+
+	var voteEntries []*governance.VoteEntry
+	for it.Seek(votesKeyFmt.Encode(id)); it.Valid(); it.Next() {
+		var proposalID uint64
+		var voter staking.Address
+		if !votesKeyFmt.Decode(it.Key(), &proposalID, &voter) {
+			break
+		}
+		if proposalID != id {
+			break
+		}
+		var vote governance.Vote
+		if err := cbor.Unmarshal(it.Value(), &vote); err != nil {
+			return nil, api.UnavailableStateError(err)
+		}
+		voteEntries = append(voteEntries, &governance.VoteEntry{
+			Voter: voter,
+			Vote:  vote,
+		})
+
+	}
+
+	return voteEntries, nil
+}
+
+func (s *ImmutableState) isProposalPendingUpgrade(ctx context.Context, proposal *governance.Proposal) (bool, error) {
+	if proposal.Content.Upgrade == nil {
+		return false, nil
+	}
+	data, err := s.is.Get(ctx, pendingUpgradesKeyFmt.Encode(proposal.Content.Upgrade.Epoch, proposal.ID))
+	return data != nil, err
+}
+
+// PendingUpgradeProposal looks up a pending upgrade proposal by its identifier.
+func (s *ImmutableState) PendingUpgradeProposal(ctx context.Context, id uint64) (*governance.UpgradeProposal, error) {
+	proposalRaw, err := s.getProposalRaw(ctx, id)
+	if err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	if proposalRaw == nil {
+		return nil, governance.ErrNoSuchProposal
+	}
+	var proposal governance.Proposal
+	if err = cbor.Unmarshal(proposalRaw, &proposal); err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	b, err := s.isProposalPendingUpgrade(ctx, &proposal)
+	if err != nil {
+		return nil, err
+	}
+	if !b {
+		return nil, governance.ErrNoSuchUpgrade
+	}
+
+	return proposal.Content.Upgrade, nil
+}
+
+// PendingUpgrades looks up all pending upgrades.
+func (s *ImmutableState) PendingUpgrades(ctx context.Context) ([]*upgrade.Descriptor, error) {
+	it := s.is.NewIterator(ctx)
+	defer it.Close()
+
+	var pendingUpgrades []*upgrade.Descriptor
+	for it.Seek(pendingUpgradesKeyFmt.Encode()); it.Valid(); it.Next() {
+		var epoch epochtime.EpochTime
+		var proposalID uint64
+		if !pendingUpgradesKeyFmt.Decode(it.Key(), &epoch, &proposalID) {
+			break
+		}
+		// Load the proposal.
+		proposal, err := s.Proposal(ctx, proposalID)
+		if err != nil {
+			return nil, err
+		}
+		if proposal.Content.Upgrade == nil {
+			// TODO: this should never happen, panic?
```

Comment:
You could wrap the error in `api.UnavailableStateError(err)` which will trigger a panic.